mlir/lib/Dialect/Ptr/IR/PtrDialect.cpp

```diff
@@ -41,6 +41,54 @@ void PtrDialect::initialize() {
       >();
 }
 
+//===----------------------------------------------------------------------===//
+// FromPtrOp
+//===----------------------------------------------------------------------===//
+
+OpFoldResult FromPtrOp::fold(FoldAdaptor adaptor) {
+  // Fold the pattern:
+  // %ptr = ptr.to_ptr %v : type -> ptr
+  // (%mda = ptr.get_metadata %v : type)?
+  // %val = ptr.from_ptr %ptr (metadata %mda)? : ptr -> type
+  // To:
+  // %val -> %v
+  auto toPtr = dyn_cast_or_null<ToPtrOp>(getPtr().getDefiningOp());
+  // Cannot fold if it's not a `to_ptr` op or the initial and final types are
+  // different.
+  if (!toPtr || toPtr.getPtr().getType() != getType())
+    return nullptr;
+  Value md = getMetadata();
+  if (!md)
+    return toPtr.getPtr();
```

Comment:
I'm not sure about this condition.
It seems to imply that all the metadata are statically encoded in the type. Couldn't there be a type where some runtime metadata are default initialized? In which case round-tripping to a `ptr` would be erasing them and folding would undo this erasing.

Reply:
The verifier of `from_ptr` will check whether `PtrLikeType` expects non-trivial metadata, and if it does, it always asks for it. The only way to not provide the metadata is if `trivial_metadata` is set, which tells the compiler it's safe to assume that all info is statically encoded in the type.

Therefore in this case, it is known from the verifier that the op can be folded.

Reply:
I don't follow what you're explaining here, can you revisit this explanation in face of not having "trivial_metadata" here anymore?